### ISSUE TITLE
git_branch extension causes high CPU usage

### ISSUE THREAD

**Issue by smartding** [tjdevries/express_line.nvim#25]
I copied the following code from another thread to try express_line, and I noticed the neovim process is constantly using around 50% CPU, and the CPU usage drops the moment I comment out the git_branch extension. Any idea why?
```lua
local bui        = require('el.builtin')
local ext        = require('el.extensions')
local sec        = require('el.sections')
local sub        = require('el.subscribe')
local col        = sec.collapse_builtin
local _space     = ' '
local _separator = ' | '
local _mode      = ext.mode
local _icon      = ext.file_icon
local _file      = bui.file_relative
local _readonly  = col{ _space, bui.readonly_list, }
local _modefied  = col{ _space, bui.modified_flag }
local _gbranch   = col{ _seperator, ext.git_branch}
local _filetype  = col{ _separator, bui.filetype_list}
local _bufloc    = col{ _separator, bui.percentage_through_window}
local _split     = sec.split
local statusline = function()
  return {
    _space,
    _mode,
    _space,
    _icon,
    _space,
    _file,
    _readonly,
    _modefied,
    _split,
    -- _gbranch,
    _filetype,
    _bufloc,
    _space,
  }
end

require('el').setup({ generator = statusline })
```

**Comment by tjdevries** [comment]
You should probably only run it on certain autocmds or on an async timer. You can see the docs for an example I think. If there isn't one, I can type more stuff here :laughing:

**Comment by smartding** [comment]
> You should probably only run it on certain autocmds or on an async timer. You can see the docs for an example I think. If there isn't one, I can type more stuff here

I read your example in README,  now I only run the git branch command on `BufEnter` event, the problem is the git branch info is sometimes not displayed after opening a file.
```lua
local bui        = require('el.builtin')
local ext        = require('el.extensions')
local sec        = require('el.sections')
local sub        = require('el.subscribe')
local col        = sec.collapse_builtin
local _space     = ' '
local _separator = ' | '
local _mode      = ext.mode
local _icon      = ext.file_icon
local _file      = bui.file
local _readonly  = col{ _space, bui.readonly_list, }
local _modefied  = col{ '[', bui.modified_flag, ']' }
local _gbranch   = sub.buf_autocmd("el_git_branch", "BufEnter", function(window,buffer) return ext.git_branch(window, buffer) end)
local _filetype  = col{ _separator, bui.filetype_list}
local _bufloc    = col{ _separator, bui.percentage_through_window}
-- local _lspstatus = require('lsp-status').status()
local _split     = sec.split
local statusline = function()
  return {
    _space,
    _mode,
    _space,
    _icon,
    _space,
    _file,
    _readonly,
    _modefied,
    _split,
    _gbranch,
    _filetype,
    _bufloc,
    -- _lspstatus,
    _space,
  }
end

require('el').setup({ generator = statusline })
```
Also could you type an example of how to integrate express_line with nvim-lua/lsp-status.nvim? I tried the config as follows, but it didn't work:
```lua
local _lspstatus = require('lsp-status').status()
local statusline = function()
  return {
    _lspstatus,
  }
end

require('el').setup({ generator = statusline })
```

**Comment by smartding** [comment]
@tjdevries any idea why the git branch info is missing for the first file I open in neovim? If I open a second file, git branch info is shown in the status line.
1. start neovim in the command line without opening any file, git branch info is shown in the status line
1. open a file, git branch info is missing
1. open a second file, git branch info is shown
1. switch back to be buffer that corresponds to the first file, git branch info is missing

**Comment by tjdevries** [comment]
Maybe should try with different autocmd, maybe WinEnter or something like that. That'd be my first guess. Maybe add buf write or something as well?

**Comment by tjdevries** [comment]
Or maybe vimenter

**Comment by smartding** [comment]
I tried WinEnter, VimEnter, BufWinEnter, BufLeave, BufNewFile, BufNew, nothing worked.

**Comment by smartding** [comment]
> Maybe should try with different autocmd, maybe WinEnter or something like that. That'd be my first guess. Maybe add buf write or something as well?

I change:
```lua
sub.buf_autocmd("el_git_branch", "BufEnter", function(window,buffer) return ext.git_branch(window, buffer) end),
```
into
```lua
sub.user_autocmd("el_git_branch", "BufEnter", function(window,buffer) return ext.git_branch(window, buffer) end),
```
and the problem is fixed.

I checked the source code of lightline, they just use plain autocmd:
```viml
autocmd WinEnter,BufEnter,BufDelete,SessionLoadPost,FileChangedShellPost * call lightline#update()
```
So which kind of autocmd should I use?

**Comment by tjdevries** [comment]
Hmmm, I'm not sure about this. I will try and debug sometime in the future. Can you ping me in a week or so if I haven't posted anything?

**Comment by smartding** [comment]
I did some digging as to why `subscribe.user_autocmd` isn't working for the first file I open.  I think I've found the reason.

It's because the `No Name` buffer and the buffer that contains the first file I open, they share the same `bufnr` 1.

When I start neovim without opening any file, the `No Name` buffer is created (`bufnr` = 1), expressline did its job, setups the status line and buffer autocmd for it.

Then I open the first file. But before the new buffer is created, the `No Name` buffer is wiped out, the buffer autocmd associated with `No Name` is also deleted.  Then the new buffer is created, for some reason it reuses the same `bufnr` 1.  expressline seems to believe status line is already generated for the buffer ( because those 2 buffers share the same window id and `bufnr` ? ), so expressline did nothing. But the buffer autocmd is already deleted at this point. Now you get a status line without a proper buf autocmd.

I wonder if there's any way to configure neovim not to reuse the `bufnr` 1, that would make the problem go away.
